libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -421,6 +436,113 @@ struct ReservedErrorSelector: public PostTypeChecker::Checker
 	}
 };
 
+class YulLValueChecker : public solidity::yul::ASTWalker
+{
+public:
+	YulLValueChecker(ASTString const& _identifierName): m_identifierName(_identifierName) {}
+	bool willBeWrittenTo() const { return m_willBeWrittenTo; }
+	using solidity::yul::ASTWalker::operator();
+	void operator()(solidity::yul::Assignment const& _assignment) override
+	{
+		if (ranges::any_of(
+			_assignment.variableNames,
+			[&](auto const& yulIdentifier) {return yulIdentifier.name.str() == m_identifierName; }
+		))
+			m_willBeWrittenTo = true;
+	}
+private:
+	ASTString const& m_identifierName;
+	bool m_willBeWrittenTo = false;
+};
+
+class LValueChecker: public ASTConstVisitor
+{
+public:
+	LValueChecker(Identifier const& _identifier):
+		m_declaration(_identifier.annotation().referencedDeclaration)
+	{}
+	bool willBeWrittenTo() const { return m_willBeWrittenTo; }
+	void endVisit(Identifier const& _identifier) override
+	{
+		solAssert(_identifier.annotation().referencedDeclaration);
+		if (
+			*_identifier.annotation().referencedDeclaration == *m_declaration &&
+			_identifier.annotation().willBeWrittenTo
+		)
+			m_willBeWrittenTo = true;
+	}
+	void endVisit(InlineAssembly const& _inlineAssembly) override
+	{
+		YulLValueChecker yulChecker{m_declaration->name()};
+		yulChecker(_inlineAssembly.operations());
+		m_willBeWrittenTo = yulChecker.willBeWrittenTo();
```

Comment:
Looks like an assembly block may reset the write flag set by anything that appears before it. Please check if this is correctly detected as a non-simple loop:

```c++
for (uint i = 0; i < 42; ++i) {
    i = type(uint).max;
    assembly {}
}
```

The correct code is probably this:
```suggestion
		m_willBeWrittenTo = m_willBeWrittenTo || yulChecker.willBeWrittenTo();
```

Or, better, only run `YulLValueChecker` when `m_willBeWrittenTo` is `false`, because otherwise there's no point - the answer is already known. You can do that in `operator()(yul::Assignment)` too to skip the `any_of` check.

Reply:
You are right. This will overwrite any previous value in the flag and also not needed when it is already true.
Fixed.